### ISSUE TITLE
Machine Learning Workflow with Feature Selection

### ISSUE THREAD

**Issue by LouiseBloch** [topepo/caret#898]
Hello,
at the moment I try to build a machine learning workflow with caret. In the process I struggle with my general understanding of how to intergrate feature selection into the workflow. 

My previous workflow, which is inspired by the `train` Function looks like this: 

&nbsp;&nbsp;Resampling
&nbsp;&nbsp;Define sets of model parameter values to evaluate
&nbsp;&nbsp;**for each** parameter set **do**
&nbsp;&nbsp;&nbsp;&nbsp;**for each** resampling iteration **do**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Hold out specific samples
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Preprocess data
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fit the model
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Predict the hold out samples
&nbsp;&nbsp;&nbsp;&nbsp;**end**
&nbsp;&nbsp;Calculate average performance across hold-outs predictions
&nbsp;&nbsp;**end**
&nbsp;&nbsp;Determine the optimal parameter set 
&nbsp;&nbsp;Fit the final model
&nbsp;&nbsp;Predict the hold-out samples
&nbsp;&nbsp;Calculate performance of hold-outs prediction

Now I would like to integrate feature selection into this workflow. My first guess was to do a filter selection after the preprocessing step. But after I read some other issues #503, #487, #439 I guessed, that this is not the way to do. I think the reason for that is that a further evaluation of the feature selection is needed. In my understanding the algorithm described in the issues is:

&nbsp;&nbsp;Resampling
&nbsp;&nbsp;**for each** resampling iteration **do**
&nbsp;&nbsp;&nbsp;&nbsp;Features Selection
&nbsp;&nbsp;&nbsp;&nbsp;Define sets of model parameter values to evaluate
&nbsp;&nbsp;&nbsp;&nbsp;**for each** parameter set **do**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**for each** resampling iteration **do**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Hold out specific samples
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Preprocess data
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fit the model
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Predict the hold out samples
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**end**
&nbsp;&nbsp;&nbsp;&nbsp;Calculate average performance across hold-outs predictions for selected features
&nbsp;&nbsp;&nbsp;&nbsp;**end**
&nbsp;&nbsp;&nbsp;&nbsp;Determine the optimal parameter set of selected features
&nbsp;&nbsp;&nbsp;&nbsp;Fit best model for selected features
&nbsp;&nbsp;&nbsp;&nbsp;Predict the hold out samples
&nbsp;&nbsp;**end**
&nbsp;&nbsp;Determine the optimal selected Features
&nbsp;&nbsp;Fit the final model
&nbsp;&nbsp;Predict the hold-out samples
&nbsp;&nbsp;Calculate performance of hold-outs prediction

I am a bit confused, so my first question is if I understood things right. If it is right I am struggeling with the outer resampling evaluation, because feature selection is performed on training and testset of the inner resampling. Isn't there a bias problem? 

Because of my confusion and because I am a beginner with machine learning it would be nice if you can post an algorithm like the one for the `train` function and maybe explain it a bit or give me a hint where I can find a good machine learning workflow which I can adopt.

Greetings, 
Louise

**Comment by topepo** [comment]
> If it is right I am struggling with the outer resampling evaluation, because feature selection is performed on training and test set of the inner resampling. Isn't there a bias problem?

In what way? 

Think of it this way: define a module that contains the parts of the modeling process. That might include preprocessing, feature filtering, model fitting, etc. 

Resampling should be executed on that module, and all parts of the module, since that is what you are trying the estimate the performance of. 

Once you are done estimating performance, the final model (i.e. the output of the model), is run on the entire training set. That model's performance was estimated using the outer resamples of the module. 

If you haven't, it might help to read sections [18](https://topepo.github.io/caret/feature-selection-overview.html) through 20 of the `caret` website.

**Comment by LouiseBloch** [comment]
Thank you very much.
